src/runtime/gdpr-tcfv2.js

```diff
@@ -0,0 +1,427 @@
+    (tcData.gdprApplies !== undefined &&
+      typeof tcData.gdprApplies !== 'boolean') ||
+    (tcData.addtlConsent !== undefined &&
+      typeof tcData.addtlConsent !== 'string')
+  ) {
+    return false;
+  }
+
+  if (!tcData.cmpStatus || tcData.cmpStatus === CmpStatus.ERROR) {
+    return false;
+  }
+
+  return true;
+};
+
+/**
+ * Returns true if the TCData object reflects user intention.
+ * @param {!./gdpr-tcfv2-constants.TCData} tcData
+ * @return {boolean}
+ */
+const isTcDataReady = tcData => {
+  if (
+    tcData.cmpStatus === CmpStatus.LOADED &&
+    (tcData.eventStatus === EventStatus.TC_LOADED ||
+      tcData.eventStatus === EventStatus.USER_ACTION_COMPLETE)
+  ) {
+    return true;
+  }
+
+  return false;
+};
+
+/**
+ * Utility class for querying the IAB TCFv2 CMP API for Transparency and Consent
+ * data.
+ */
+export class TcfV2 {
+  /**
+   * @param {!Window} win The frame which the vendor is running in.
+   * @param {number=} timeoutMs How long to wait for the callback before
+   *     erroring out.
+   */
+  constructor(win, timeoutMs = DEFAULT_ERROR_TIMEOUT_MS) {
+    /**
+     * The frame which this code is running in.
+     * @private {!Window}
+     */
+    this.window_ = win;
+
+    /**
+     * The frame where the CMP is located.
+     * @private {?Window}
+     */
+    this.cmpFrameWin_ = null;
+
+    /**
+     * Store callbacks until the CMP replies.
+     * @private {!Object<number, function(!./gdpr-tcfv2-constants.TCData, boolean): undefined>}
+     */
+    this.pendingCalls_ = {};
+
+    /**
+     * A simple counter to disambiguate multiple callback messages from outer
+     * frames.
+     * @private {number}
+     */
+    this.callbackIndex_ = 0;
+
+    /**
+     * How long to wait for the callback before erroring out.
+     * @private @const {number}
+     */
+    this.timeoutMs_ = timeoutMs;
+
+    /**
+     * Postmessage listener bound to this object.
+     * @private {?function(!Object):undefined}
+     */
+    this.boundHandler_ = null;
+  }
+
+  /**
+   * Eliminates all traces of this component from the page
+   */
+  disposeInternal() {
+    this.pendingCalls_ = {};
+    if (this.boundHandler_) {
+      unregisterEventHandler(this.window_, 'message', this.boundHandler_);
+      delete this.boundHandler_;
+    }
+    delete this.pendingCalls_;
+    delete this.window_;
+    delete this.cmpFrameWin_;
+  }
+
+  /**
+   * @return {boolean}
+   */
+  isCmpAvailable() {
+    return isFunction(this.window_['__tcfapi']) || this.findCmpFrame() != null;
+  }
+
+  /**
+   * Invokes the callback with whatever is available in the current state of the
+   * CMP. If CMP is on the same domain, it is expected to be synchronous.
+   * @param {function(!./gdpr-tcfv2-constants.TCData): undefined} clientCallback
+   * @param {!Array<string>|undefined} vendorIds Which vendors to lookup, or
+   *     undefined for all vendors.
+   */
+  getTCData(clientCallback, vendorIds) {
```

Comment:
I modified this function, I am considered about the possibility of an infinitely loop.  Is TC_STRING_UNAVAILABLE something you just wait until its is ready for or is it a permanent state?